### ISSUE TITLE
LZ4 support?

### ISSUE THREAD

**Issue by subes** [pcmind/leveldb#18]
Hi, I really love your fork of LevelDB. I tried a few months ago myself but could not get it adapted to use: https://github.com/lz4/lz4-java 
Though I guess this could lead to significantly better performance than with snappy?

**Comment by pcmind** [comment]
Apparently LZ4 have a better read performance, so depending on the usecase and data, it could lead to a better performance.
I'll give it a try.

**Comment by subes** [comment]
Very nice! Here a pointer for how to use it: https://github.com/invesdwin/invesdwin-context/blob/ca3e846b67bc31df7eb95e91dac8328c3ab6c08d/invesdwin-context-parent/invesdwin-context-integration/src/main/java/de/invesdwin/context/integration/streams/compressor/lz4/LZ4Streams.java
See compress/decompress methods at the bottom.

Also we could try both LZ4Fast and LZ4HC dependending on the use case. There are significant differences between the algos:
```
 ATimeSeriesDB (None)      Writes (Append):      1,033.68/ms  => ~5.3 times faster than LevelDB-Java (with Disabled Compression)
 ATimeSeriesDB (High)      Writes (Append):      5,819.10/ms  => ~29.8 times faster than LevelDB-Java (with LZ4 High Compression)
 ATimeSeriesDB (Fast)      Writes (Append):     21,095.74/ms  => ~108 times faster than LevelDB-Java (with LZ4 Fast Compression)
 ATimeSeriesDB (None)       Reads (Iterator):   19,638.26/ms  => ~18.9 times faster than LevelDB-Java
 ATimeSeriesDB (High)       Reads (Iterator):   32,089.34/ms  => ~30.9 times faster than LevelDB-Java
 ATimeSeriesDB (Fast)       Reads (Iterator):   32,124.39/ms  => ~30.95 times faster than LevelDB-Java
```
RocksDB also supports LZ4 and LZ4HC:
```
package org.rocksdb;

/**
 * Enum CompressionType
 *
 * <p>DB contents are stored in a set of blocks, each of which holds a
 * sequence of key,value pairs. Each block may be compressed before
 * being stored in a file. The following enum describes which
 * compression method (if any) is used to compress a block.</p>
 */
public enum CompressionType {
  NO_COMPRESSION((byte) 0x0, null),
  SNAPPY_COMPRESSION((byte) 0x1, "snappy"),
  ZLIB_COMPRESSION((byte) 0x2, "z"),
  BZLIB2_COMPRESSION((byte) 0x3, "bzip2"),
  LZ4_COMPRESSION((byte) 0x4, "lz4"),
  LZ4HC_COMPRESSION((byte) 0x5, "lz4hc"),
  XPRESS_COMPRESSION((byte) 0x6, "xpress"),
  ZSTD_COMPRESSION((byte)0x7, "zstd"),
  DISABLE_COMPRESSION_OPTION((byte)0x7F, null);
```
Though anything going through JNI suffers a lot from performance penalties as visible here:
https://github.com/invesdwin/invesdwin-context-persistence/blob/master/README.md#timeseries-module

I think if we stay within java this should work a lot better.

**Comment by pcmind** [comment]
Thank you for the references. I'll implent it using the lz4-java as suggested and use same id as rocksdb.

**Comment by pcmind** [comment]
Are you available to test the changes I made in the new lz4compression branch?

**Comment by subes** [comment]
Yes, I will test it this evening.

**Comment by subes** [comment]
I did do these tests on my notebook, so there might be differences also due to thermal downclocking. Though I ran each test multiple times to validate it does not fluctuate too much.

`i9 9900K, SSD, Java 16`

DB-Size 10 million small entries:

**PutBatch**
```
265.71/ms    LZ4HC Native
267.07/ms    LZ4HC Java
267.37/ms    LZ4 Native
271.54/ms    Snappy iq80
273.29/ms    Snappy Xerial
273.73/ms    LZ4 Java
288.12/ms    None
```

**Iterate**
```
1926.60/ms    LZ4 Native
2152.99/ms    Snappy iq80
2158.85/ms    LZ4 Java
2166.80/ms    LZ4HC Java
2231.35/ms    None
2303.62/ms    Snappy Xerial
2386.07/ms    LZ4HC Native
```

DB-Size 100k small entries:

**Get**
```
291.76/ms    LZ4HC Native            
293.26/ms    Snappy Xerial
297.09/ms    LZ4 Native
299.63/ms    None
301.87/ms    LZ4HC Java
304.40/ms    LZ4 Java
304.92/ms    Snappy iq80
```

**GetLatest**
```
190.21/ms    LZ4HC Native
192.89/ms    Snappy Xerial     
194.19/ms    LZ4HC Java
194.42/ms    LZ4 Native
195.67/ms    None               
195.78/ms    LZ4 Java
197.56/ms    Snappy iq80    
```

So in all I don't see too much difference in the various compression algos. I think with larger databases this might be different. Though for my use case I am storing mainly small data entries for indexing purposes.

Nonetheless, the LZ4 implementation looks great. I guess I would do more extensive testing at a later point, but I would go for LZ4 HC for now. So good work there!

-----------------------------------------------------------

I guess the compression algorithm does not make much difference because the bottlenecks in leveldb are somewhere else.
On that note I optimized away some unnecessary byte[] copies by exposing Slice methods directly in DBImpl and making the DBComparator zero-copy. This almost doubled the performance for iteration from ~1000/ms to ~2000/ms. So unnecessary copying seems to be a more significant performance hog. 
Can we incorporate the extensions to the three internal leveldb classes I made for this here (https://github.com/criccomini/ezdb/tree/master/ezdb-leveldb/src/main/java/org/iq80/leveldb)? I guess we could make the changes backwards compatible by just allowing subclasses of DBImpl to override implementations for UserComparator via a protected factory method. The methods that operate on Slices could be added from the outside.
I prepared in "ExtensibleDBImpl" a version that is designed for extensibility (though more private methods could be made protected). I only modified all fields being protected and making the methods protected that I need for my extensions.
=> I have extracted this into a different ticket: https://github.com/pcmind/leveldb/issues/19

Maybe one would also have to disable the leveldb cache option to see the differences between the compression algos better. But that would be a more synthetic test. Thus I think testing with larger data entries would be the way to go.
Other things that could be tested would be if larger block sizes show more significant differences here, since 4kb is a bit small for any compression algorithm (i assume the compression operates on these 4kb blocks currently).
=> No actions required here, just some thoughts on better performance testing of compression algos.

Also I think there could be a lot to be gained if Slice would optionally use ByteBuffer's and reference direct memory where possible. Maybe we could support multiple implementations of Slice for zero copy optimizations where appropriate. I have not looked deeply into this option yet. But with less copying, maybe the compression algorithm will make a bigger impact then. I currently smell that the garbage collector could be causing the more or less random results. Though maybe making Slices reference direct memory would not work because of parallel compaction processes moving the segments around.
=> Could be put into a separate ticket as an idea to further improve performance, though I leave that to you to decide if this is feasible and worth the effort.
